boards: samples: Add multicore support for NXP mcxn947.

TomasGalbickaNXP opened the pull request:
This PR adds multicore support for frdm_mcxn947. The cpu1 is booted from flash.

Samples supported by this commit:
- samples/drivers/mbox
- samples/drivers/mbox_data
- samples/subsys/ipc/ipc_service
- samples/subsys/ipc/openamp

The frdm_mcxn947 supports the mbox driver.

danieldegrasse commented on samples/drivers/mbox/remote/boards/frdm_mcxn947_mcxn947_cpu1.conf:
```diff
@@ -0,0 +1,4 @@
+CONFIG_MBOX_NXP_IMX_MU=y
+CONFIG_BUILD_OUTPUT_INFO_HEADER=y
```
Typically this only needs to be set when we are copying the secondary image into RAM (and therefore need to know the VMA the image should be loaded at). Have you tried building the sample without this Kconfig set?

danieldegrasse reviewed (commented):
This is great to see, thanks. I think we should split the commits here out a bit, here's a suggested set of commits:
- commit modifying MBOX and IPM drivers for MCX support, and adding the DTS entries for the MBOX peripheral
- commit adding support for booting the second core at the SOC level (IE the SYS_INIT routine)
- commit adding board definitions and CMake changes to the board to enable dual core support
- commit adding support for each dual core sample. We could also this into commits for each sample, but I don't really think that's needed.

DerekSnell reviewed (changes requested):
Hi @TomasGalbickaNXP ,
Building for `cpu1` fails for me because of this [typo](https://github.com/zephyrproject-rtos/hal_nxp/blob/c42d70dba969de09c0e43d3b26a3ffcb015f6f33/mcux/mcux-sdk/devices/MCXN947/system_MCXN947_cm33_core1.c#L88) in the hal_nxp: `#enidf` instead of `#endif`.  It appears we also need to update the HAL to resolve this, and update the West manifest in your PR.

To replicate, I see this error building the `hello_world` sample with this command:
```
west build -b frdm_mcxn947//cpu1 samples/hello_world/ --pristine
```
The error:
```
/home/snellda/zNXPmicro/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/devices/MCXN947/system_MCXN947_cm33_core1.c: In function 'SystemInit':
/home/snellda/zNXPmicro/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/devices/MCXN947/system_MCXN947_cm33_core1.c:80: error: unterminated #if
   80 | #if !defined(__ZEPHYR__)
      | 
/home/snellda/zNXPmicro/zephyrproject/modules/hal/nxp/mcux/mcux-sdk/devices/MCXN947/system_MCXN947_cm33_core1.c:78:3: error: expected declaration or statement at end of input
   78 |   SYSCON->NVM_CTRL &= ~SYSCON_NVM_CTRL_DIS_MBECC_ERR_DATA_MASK; /* enables bus error on multi-bit ECC error for data */
```
I initially tried building the OpenAMP sample, and had the same error:
```
west build -b frdm_mcxn947//cpu0 --sysbuild samples/subsys/ipc/openamp --pristine
```

DerekSnell reviewed (changes requested):
Also blocking until update the [board doc page](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/nxp/frdm_mcxn947/doc/index.rst).  Similar to Daniel's request, we need to add details about cpu1 target.  
- Update the [Targets available](https://docs.zephyrproject.org/latest/boards/nxp/frdm_mcxn947/doc/index.html#targets-available) section for cpu1.  See the [LPC55S69](https://docs.zephyrproject.org/latest/boards/nxp/lpcxpresso55s69/doc/index.html#targets-available) page
- Add an example `west build` command with cpu1
- Add a section "Building a dual-core image", see [LPC55S69](https://docs.zephyrproject.org/latest/boards/nxp/lpcxpresso55s69/doc/index.html#building-a-dual-core-image).

TomasGalbickaNXP committed:
dts: drivers: Add DTS MBOX entry for NXP MCXN947

This commit adds MBOX device tree entry for MCXN947.
Adds support for MCXN in NXP ipm and mbox drivers.

Signed-off-by: Tomas Galbicka <tomas.galbicka@nxp.com>

TomasGalbickaNXP committed:
boards: Add multicore support on CPU1 for MCXN947

This commit adds multicore support for MCXN947.
It enables the secondary core CPU1 to boot from flash.

Signed-off-by: Tomas Galbicka <tomas.galbicka@nxp.com>

TomasGalbickaNXP commented on samples/drivers/mbox_data/remote/boards/frdm_mcxn947_mcxn947_cpu1.conf:
```diff
@@ -0,0 +1,4 @@
+CONFIG_MBOX_NXP_IMX_MU=y
+CONFIG_BUILD_OUTPUT_INFO_HEADER=y
```
Removed it is not required.

TomasGalbickaNXP commented on samples/drivers/mbox/remote/boards/frdm_mcxn947_mcxn947_cpu1.conf:
```diff
@@ -0,0 +1,4 @@
+CONFIG_MBOX_NXP_IMX_MU=y
```
Removed it is not required.

TomasGalbickaNXP commented on boards/nxp/frdm_mcxn947/frdm_mcxn947.dtsi:
```diff
@@ -132,6 +132,17 @@ nxp_8080_touch_panel_i2c: &flexcomm2_lpi2c2 {
+		compatible = "mmio-sram";
+		reg = <0x20050000 DT_SIZE_K(64)>;
+	};
+	sram2: memory@20060000 {
```
Good suggestion I have changed so ram naming correlates with TRM.

TomasGalbickaNXP commented on boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu1.yaml:
```diff
@@ -0,0 +1,22 @@
+name: NXP FRDM MCXN947 (CPU1)
+type: mcu
+arch: arm
+ram: 320
```
Size was corrected, Thanks.

TomasGalbickaNXP commented on boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu1.yaml:
```diff
@@ -0,0 +1,22 @@
+type: mcu
+arch: arm
+ram: 320
+flash: 2048
```
Size was corrected, Thanks.

TomasGalbickaNXP commented on boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu1.dtsi:
```diff
@@ -0,0 +1,29 @@
+	};
+
+	chosen {
+		zephyr,sram = &sram1;
```
This one should be ramg, this is ram partition for data. (changed)
Booting is done from flash, so no RAM boot partition required here.
The sramh is reserved for Shared Memory for OpenAMP communication.

TomasGalbickaNXP commented on samples/drivers/mbox/remote/boards/frdm_mcxn947_mcxn947_cpu1.conf:
```diff
@@ -0,0 +1,4 @@
+CONFIG_MBOX_NXP_IMX_MU=y
+CONFIG_BUILD_OUTPUT_INFO_HEADER=y
+CONFIG_BUILD_OUTPUT_HEX=y
```
These are not required here, no code relocation. (removed)

TomasGalbickaNXP commented on samples/subsys/ipc/openamp/remote/boards/frdm_mcxn947_mcxn947_cpu1.overlay:
```diff
@@ -0,0 +1,28 @@
+		 * shared memory reserved for the inter-processor communication
+		 */
+
+		zephyr,ipc_shm = &sram2;
```
changed accordingly. Thanks.

TomasGalbickaNXP commented:
Hi @DerekSnell regarding https://github.com/zephyrproject-rtos/zephyr/pull/81882#pullrequestreview-2459381107
Yes it is know issue, and PR is created to nxp_hal, waiting for a merge.

danieldegrasse commented on boards/nxp/frdm_mcxn947/doc/index.rst:
```diff
@@ -259,6 +310,12 @@ should see the following message in the terminal:
+---------------------------
+
+For dual core builds, the secondary core should be placed into a loop,
+then a debugger can be attached (see `AN13264`_, section 4.2.3 for more information)
```
Sorry, should have been more clear here. AN13264 is specifically about the RT1170- do we have an appnote like this for the MCXN? If not, we can probably link to this appnote and say "using a technique similar to that described for the RT1170 in AN13264"

danieldegrasse commented on west.yml:
Place the commit making the manifest change first, since the other features are dependent upon it.

TomasGalbickaNXP replied:
Okay Thank you I have moved commit with manifest change as first one.

danieldegrasse commented:
@DerekSnell should we create a multi-column table indicating which features are supported on the M33 secondary core, like we do for the RT1170? Or no need, since both cores are an M33?

DerekSnell commented:
> @DerekSnell should we create a multi-column table indicating which features are supported on the M33 secondary core, like we do for the RT1170? Or no need, since both cores are an M33?

Thanks for checking, @danieldegrasse .  But I do not think we need to complicate the Supported Features table with this.  The RT1170 needs that table because the memory maps are different between the two cores.  But that is not the case with this MCX device, and similar multicore devices like LPC5569 did not need these columns.  Thanks

DerekSnell commented on boards/nxp/frdm_mcxn947/doc/index.rst:
```diff
@@ -204,6 +236,25 @@ see the following message in the terminal:
+The dual-core samples are run using ``frdm_mcxn947/mcxn947/cpu0`` target.
+
+Images built for ``frdm_mcxn947/mcxn947/cpu1`` will be loaded from flash
+and executed on the second core when ``SECOND_CORE_MCUX`` is selected.
```
```suggestion
and executed on the second core when ``CONFIG_SECOND_CORE_MCUX`` is selected.
```

kartben replied:
even better is the following :)

```suggestion
and executed on the second core when :kconfig:option:`CONFIG_SECOND_CORE_MCUX` is selected.
```

TomasGalbickaNXP replied:
Thanks for great tip, updated.

DerekSnell reviewed (dismissed):
Thank you @TomasGalbickaNXP for addressing all my requests.

These samples all build and run for me.  I found some minor improvements in the board doc in case you push any updates.  But I am approving.  Thank you

TomasGalbickaNXP commented on boards/nxp/frdm_mcxn947/doc/index.rst:
```diff
@@ -101,14 +101,46 @@ The FRDM-MCXN947 board configuration supports the following hardware features:
++-----------+------------------+----------------------+
+| Core      | Boot Address     | Comment              |
++===========+==================+======================+
+| CPU0      | 0x00010000[992K] | primary core flash   |
```
updated

danieldegrasse reviewed (commented):
I think we need to create a _defconfig for this target: `frdm_mcxn947_mcxn947_cpu1_defconfig`. Currently when building dual core images, features like GPIO and console support aren't enabled for the second core. Is this intentional?


There are also a few Kconfig warnings when targeting the second core:
```
warning: SOC_RESET_HOOK (defined at soc/xlnx/zynq7000/Kconfig.defconfig:23, kernel/Kconfig.init:9)
was assigned the value 'n' but got the value 'y'. See
http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_SOC_RESET_HOOK and/or look up
SOC_RESET_HOOK in the menuconfig/guiconfig interface. The Application Development Primer, Setting
Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be helpful
too.


warning: SEGGER_RTT_BUFFER_SIZE_UP (defined at modules/segger/Kconfig:36, modules/segger/Kconfig:36)
was assigned the value '4096' but got the value ''. Check these unsatisfied dependencies:
(USE_SEGGER_RTT || (USE_SEGGER_RTT && 0)) (=n). See
http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_SEGGER_RTT_BUFFER_SIZE_UP and/or look up
SEGGER_RTT_BUFFER_SIZE_UP in the menuconfig/guiconfig interface. The Application Development Primer,
Setting Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be
helpful too.


warning: PRINTK (defined at subsys/debug/Kconfig:220) was assigned the value 'n' but got the value
'y'. See http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_PRINTK and/or look up PRINTK in
the menuconfig/guiconfig interface. The Application Development Primer, Setting Configuration
Values, and Kconfig - Tips and Best Practices sections of the manual might be helpful too.
```

I'm slightly unclear how the image is being loaded. I see that the second core image has a `zephyr,flash` node of `slot1_partition`- this gives the image a load address of `0x10_a000`. This is in the nonsecure flash region.

What I suspect might be happening is that we are loading the image to the nonsecure flash *address*, but since that region is mirrored in the secure flash space, no issue appears.

I think the right way to do this is to set the `zephyr,flash` node to `&flash`, just like CPU0. However, in the `_defconfig` for the CPU1, we should set `CONFIG_USE_DT_CODE_PARTITION=y`. Then, the image will be linked into the code partition like we expect, and will be loaded into the secure address space at `0x1010_a000` as expected

mmahadevan108 commented:
Please update the `west.yml` file

mmahadevan108 commented on drivers/ipm/ipm_mcux.c:
```diff
@@ -20,8 +20,14 @@
+defined(MCXN947_cm33_core0_SERIES) || defined(MCXN947_cm33_core1_SERIES) || \
+defined(MCXN946_cm33_core0_SERIES) || defined(MCXN946_cm33_core1_SERIES) || \
+defined(MCXN547_cm33_core0_SERIES) || defined(MCXN547_cm33_core1_SERIES) || \
+defined(MCXN546_cm33_core0_SERIES) || defined(MCXN546_cm33_core1_SERIES))
```
Can we use `CONFIG_SOC_SERIES_MCXN` instead of the above list?

TomasGalbickaNXP replied:
Yes we can, Thanks for the tip.

TomasGalbickaNXP commented on boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu1.yaml:
```diff
@@ -0,0 +1,20 @@
+type: mcu
+arch: arm
+ram: 64
+flash: 992
```
updated, Thanks.

butok commented on boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu0.dtsi:
```diff
@@ -27,6 +27,7 @@
 		zephyr,console = &flexcomm4_lpuart4;
 		zephyr,shell-uart = &flexcomm4_lpuart4;
 		zephyr,canbus = &flexcan0;
+		zephyr,code-cpu1-partition = &slot1_partition;
```
Here you are trying to use MCUBoot slot1 partition for Core1. 
But the current partition table was defined for MCUBoot - not for multi-core use-case.
Multi-core have to define own variant of partition table.

TomasGalbickaNXP replied:
I have created separate 200K partition for cpu1 image, would this be better option?
https://github.com/zephyrproject-rtos/zephyr/pull/81882/files#diff-62c4d7185a8523df62bb0a05951062b2770017f591408a07a0f5fb4777e20343R167

butok replied:
> @butok , FYI. It has been common practice to place the code for the secondary core in the `slot1_partition`. LPC MCUs have used [zephyr,code-cpu1-partition](https://github.com/search?q=repo%3Azephyrproject-rtos%2Fzephyr%20zephyr%2Ccode-cpu1-partition&type=code) to do this, and RT11xx has used [nxp,m4-partition](https://github.com/search?q=repo%3Azephyrproject-rtos%2Fzephyr+m4-partition&type=code) to do the same. I assume it has been an easy option to re-use an existing flash partition that does not conflict with the primary core.

We have one common MCU-Boot specific partition setup that we try to use for all use cases.
The best solution would be something like the Nordic's Partition Manager (https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/scripts/partition_manager/partition_manager.html), but they don't plan to upstream it.

butok commented on boards/nxp/frdm_mcxn947/frdm_mcxn947.dtsi:
```diff
@@ -147,11 +158,15 @@ nxp_8080_touch_panel_i2c: &flexcomm2_lpi2c2 {
-			reg = <0x0010A000 DT_SIZE_K(984)>;
+			reg = <0x000F1000 DT_SIZE_K(884)>;
+		};
+		cpu1_partition: partition@1CE000 {
```
OK. What about cpu0_partition?

TomasGalbickaNXP commented on boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu1.dtsi:
I just left the structure as if was for cpu0. That could be possible to merge into `cpu1.dts`.
Maybe Question here is if we would like to also support cpu1 from qspi?

danieldegrasse replied:
Perhaps in the future, at which point we could split this back out. For now, we should merge them

TomasGalbickaNXP commented on boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu1_defconfig:
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+CONFIG_GPIO=y
```
OK That is a good Idea, I see there is also second uart out in board ref. manual. I will enable this.

TomasGalbickaNXP commented on soc/nxp/mcx/mcxn/Kconfig:
```diff
@@ -7,8 +7,6 @@ config SOC_SERIES_MCXN
 	select HAS_MCUX_FLEXCOMM
-	select CPU_CORTEX_M_HAS_SYSTICK
-	select CPU_CORTEX_M_HAS_DWT
 	select SOC_RESET_HOOK
```
After checking it does make sense to run it only on primary core. SystemInit is almost the same for cpu0 and cpu1, and only cpu0 can start cpu1 so it should be initialized only once.
I will update.

danieldegrasse commented:
At least for me in order to attach with a debugger when running multicore samples, Linkserver needs to be launched with the --all argument for the cores to use. I think we need a change like this in the `board.cmake`:

```diff
diff --git a/boards/nxp/frdm_mcxn947/board.cmake b/boards/nxp/frdm_mcxn947/board.cmake
index e7db5aa75df..10f176e8af1 100644
--- a/boards/nxp/frdm_mcxn947/board.cmake
+++ b/boards/nxp/frdm_mcxn947/board.cmake
@@ -7,7 +7,11 @@
 if(CONFIG_SOC_MCXN947_CPU0 OR CONFIG_SECOND_CORE_MCUX)
   board_runner_args(jlink "--device=MCXN947_M33_0" "--reset-after-load")
   board_runner_args(linkserver  "--device=MCXN947:FRDM-MCXN947")
-  board_runner_args(linkserver  "--core=cm33_core0")
+  if (CONFIG_SECOND_CORE_MCUX)
+    board_runner_args(linkserver  "--core=all")
+  else()
+    board_runner_args(linkserver  "--core=cm33_core0")
+  endif()
   board_runner_args(linkserver  "--override=/device/memory/1/flash-driver=MCXN9xx_S.cfx")
   board_runner_args(linkserver  "--override=/device/memory/1/location=0x10000000")
   # Linkserver v1.4.85 and earlier do not include the secure regions in the
```

danieldegrasse commented on boards/nxp/frdm_mcxn947/doc/index.rst:
nit- doc changes should be in prior commit adding board support

TomasGalbickaNXP replied:
fixed.

TomasGalbickaNXP committed:
samples: enable multicore samples for MCXN947 board

This commit adds support for these multicore samples:
- samples/drivers/mbox
- samples/drivers/mbox_data
- samples/subsys/ipc/ipc_service/static_vrings
- samples/subsys/ipc/openamp

Signed-off-by: Tomas Galbicka <tomas.galbicka@nxp.com>

danieldegrasse commented:
@TomasGalbickaNXP please rebase when you get a chance

decsny commented on boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu1.dts:
```diff
@@ -0,0 +1,80 @@
+	status = "okay";
+};
+
+&w25q64jvssiq {
```
aren't some of these devices being status=okay going to conflict with what devices are status=okay for cpu0 ? I don't think most of these drivers have been written with the intention of two different cores doing i/o doing their own i/o and management of the peripherals

TomasGalbickaNXP replied:
I think I was adding them because some tests here in CI failed.
Because they try to run some flash example on standalone CPU1.
I removed them, let see what CI will tell.

TomasGalbickaNXP replied:
Yes the flash is required, otherwise the test `frdm_mcxn947/mcxn947/cpu1 drivers.flash.common.default` fails.

decsny replied:
that's because the Devicetree has a "fixed-partitions" node with a storage partition on the external winbond flash node, is this something we need to support having enabled on both cores simultaneously?  my opinion is to move this definition out of the common board .dtsi and into the cpu0 specific file

TomasGalbickaNXP replied:
Thanks for the tip here, I moved the storage partition to cpu0 only and CI passes now!

decsny commented on boards/nxp/frdm_mcxn947/board.cmake:
```diff
@@ -4,10 +4,14 @@
   board_runner_args(jlink "--device=MCXN947_M33_0" "--reset-after-load")
   board_runner_args(linkserver  "--device=MCXN947:FRDM-MCXN947")
-  board_runner_args(linkserver  "--core=cm33_core0")
+  if (CONFIG_SECOND_CORE_MCUX)
```
i have been told we should put no space after if and between parentheses in cmake ie `if()` not `if ()`

decsny replied:
@nordicjm blocked one of my PR about this today is why I mention it

nordicjm replied:
> I don't think this is a rule. There are hundreds of examples of this format in the cmake hierarchy. Although I would say that there should be consistency in the same file.

Something being added as wrong doesn't mean it can be treated as right, there should be no spaces between the opcode and bracket in cmake

nashif replied:
time to fix the tree (quick grep shows at least 347 occurances of `if ()`, enable cmakelint in compliance check, sit back and watch people fighting bots :)

nashif replied:
for reference:

https://pypi.org/project/cmakelint/


and it does catch this case:

```CMakeLists.txt:320: Extra spaces between 'if' and its () [whitespace/extra]```

nordicjm replied:
This is planned though using cmake-format as per https://github.com/zephyrproject-rtos/zephyr/pull/82839

TomasGalbickaNXP commented on drivers/mbox/mbox_nxp_mailbox.c:
```diff
@@ -21,8 +21,11 @@ LOG_MODULE_REGISTER(nxp_mbox_mailbox);
-#if (defined(LPC55S69_cm33_core0_SERIES) || defined(LPC55S69_cm33_core1_SERIES))
-#ifdef LPC55S69_cm33_core0_SERIES
+#if (defined(LPC55S69_cm33_core0_SERIES) || defined(LPC55S69_cm33_core1_SERIES) ||                 \
+defined(CONFIG_SOC_SERIES_MCXN))
```
Yes, this series all using the mailbox IP.

TomasGalbickaNXP commented on drivers/ipm/ipm_mcux.c:
```diff
@@ -20,8 +20,11 @@
-#if (defined(LPC55S69_cm33_core0_SERIES) || defined(LPC55S69_cm33_core1_SERIES))
-#ifdef LPC55S69_cm33_core0_SERIES
+#if (defined(LPC55S69_cm33_core0_SERIES) || defined(LPC55S69_cm33_core1_SERIES) || \
+defined(CONFIG_SOC_SERIES_MCXN))
```
Yes, this series all using the mailbox IP.

TomasGalbickaNXP commented on soc/nxp/mcx/mcxn/soc.c:
```diff
@@ -38,3 +38,32 @@ void soc_reset_hook(void)
+ *
+ * @retval 0 on success.
+ *
+ */
```
Removed

TomasGalbickaNXP commented on soc/nxp/mcx/mcxn/Kconfig:
```diff
@@ -7,18 +7,23 @@ config SOC_SERIES_MCXN
+
+config SOC_MCXN947_CPU1
+	select CPU_CORTEX_M33
+	select HAS_MCUX_CACHE
```
Correct, removed cache.

TomasGalbickaNXP commented on soc/nxp/mcx/mcxn/Kconfig:
```diff
@@ -7,18 +7,23 @@ config SOC_SERIES_MCXN
+	select SOC_RESET_HOOK
 	select ARM_TRUSTZONE_M
 	select HAS_MCUX_CACHE
+	select CPU_CORTEX_M_HAS_SYSTICK
```
Yes correct, Thanks, Added to CPU1 core.

TomasGalbickaNXP commented on soc/nxp/mcx/mcxn/Kconfig:
```diff
@@ -7,18 +7,23 @@ config SOC_SERIES_MCXN
 	select ARM_TRUSTZONE_M
 	select HAS_MCUX_CACHE
+	select CPU_CORTEX_M_HAS_SYSTICK
+	select CPU_CORTEX_M_HAS_DWT
```
Yes correct, Thanks, Added to CPU1 core.

TomasGalbickaNXP commented on soc/nxp/mcx/mcxn/soc.c:
```diff
@@ -26,9 +26,9 @@ void soc_reset_hook(void)
-		     (DT_NODE_HAS_COMPAT(n, nxp_lpi2c) == 0),			\
+#define FLEXCOMM_CHECK_2(n)                                                                        \
+	BUILD_ASSERT((DT_NODE_HAS_COMPAT(n, nxp_kinetis_lpuart) == 0) &&                           \
+			     (DT_NODE_HAS_COMPAT(n, nxp_imx_lpi2c) == 0),                          \
```
oh, thank you for a catch here. added back nxp_lpi2c. propably rebasing issue.

TomasGalbickaNXP commented on dts/arm/nxp/nxp_mcxn94x_common.dtsi:
```diff
@@ -534,6 +534,15 @@
 
+	mbox:mbox@b2000 {
+		compatible = "nxp,mbox-mailbox";
+		reg = <0xb2000 0xEC>;
```
fixed.

TomasGalbickaNXP commented on dts/arm/nxp/nxp_mcxn94x_common.dtsi:
```diff
@@ -534,6 +534,15 @@
 		};
 	};
 
+	mbox:mbox@b2000 {
```
fixed.

TomasGalbickaNXP commented on samples/drivers/mbox_data/remote/boards/frdm_mcxn947_mcxn947_cpu1.overlay:
```diff
@@ -0,0 +1,14 @@
+		mboxes = <&mbox 0>, <&mbox 1>;
+		mbox-names = "tx", "rx";
+	};
+
```
Fixed.

TomasGalbickaNXP commented on soc/nxp/mcx/mcxn/Kconfig:
```diff
@@ -7,18 +7,23 @@ config SOC_SERIES_MCXN
 	select HAS_MCUX_CACHE
+	select CPU_CORTEX_M_HAS_SYSTICK
+	select CPU_CORTEX_M_HAS_DWT
+
```
Fixed.

kartben commented on boards/nxp/frdm_mcxn947/doc/index.rst:
```diff
@@ -204,6 +240,25 @@ see the following message in the terminal:
+and executed on the second core when :kconfig:option:`CONFIG_SECOND_CORE_MCUX` is selected.
+
+For an example of building for both cores with System Build, see
+``samples/subsys/ipc/ipc_service/static_vrings``
```
Yes, that should work (it should even automatically check that the file/directory exists so yeah, it's recommended to use it!)

TomasGalbickaNXP replied:
Great! Thank you, updated!

decsny commented:
see comment above about flash ,  I think still we should have minimal things enabled on cpu1

kartben commented on boards/nxp/frdm_mcxn947/doc/index.rst:
```diff
@@ -101,14 +101,46 @@ The FRDM-MCXN947 board configuration supports the following hardware features:
+- ``samples/subsys/ipc/ipc_service/static_vrings``
+- ``samples/subsys/ipc/openamp``
+- ``samples/drivers/mbox``
+- ``samples/drivers/mbox_data``
```
Samples should actually be cross-referenced using the `:zephyr:code-sample:` role, using the identifier defined in their respective READMEs. 

Ex:

```
- :zephyr:code-sample:`ipc-static-vrings`
- ...
```

It's fine with me if you don't feel like changing it again at this stage :)

hakehuang commented on boards/nxp/frdm_mcxn947/board.cmake:
```diff
@@ -4,10 +4,14 @@
+  if(CONFIG_SECOND_CORE_MCUX)
+    board_runner_args(linkserver  "--core=all")
+  else()
+    board_runner_args(linkserver  "--core=cm33_core0")
```
this not work in frdm_mcxn236
